lib/process_settings/abstract_monitor.rb

```diff
@@ -94,10 +96,24 @@ def static_context=(context)
     def targeted_value(*path, dynamic_context:, required: true)
       # Merging the static context in is necessary to make sure that the static context isn't shifting
       # this can be rather costly to do every time if the dynamic context is not changing
-      # TODO: Warn in the case where dynamic context was attempting to change a static value
-      # TODO: Cache the last used dynamic context as a potential optimization to avoid unnecessary deep merges
-      # TECH-4402 was created to address these todos
-      full_context = dynamic_context.deep_merge(static_context)
+
+      # Warn in the case where dynamic context was attempting to change a static value
+      dynamic_context.each do |key, value|
+        if static_context.key? key
+          warn("WARNING: Dynamic key: '#{key}' overlaps with Static key!")
+          break
+        end
+      end
+
+      # To avoid unnecessary deep merge store deep_merges in class variable hash and lookup before deep_merge
+      key = static_context.to_s + dynamic_context.to_s
```

Comment:
We should put a size limit on this hash so that we don't endlessly bloat the memory.  A very crude way to do this would be to reset the hash every time it reaches a certain size.  Let's also add some logging to this if statement so that we can track cache hits and misses.

Reply:
Also the key only needs to be the `dynamic_context` since the `static_context` is restricted from changing. And in ruby, hash keys can be objects, so there no need to `to_s` the key

Reply:
@jebentier What would an appropriate hash size limit be?